### PR TITLE
Use nullable references in security rules

### LINKED ISSUE #3069: Fix TODO comments to enable nullable reference types in security analyzers
#3021 enabled nullable reference types for the repo, but a bunch of security analyzers and analysis files were opted out. @dotpaul, we need to fix the following two buckets of suppressions:

1.
```csharp
// TODO(dotpaul): Enable nullable analysis.
#nullable disable
```

2. 
```csharp
// TODO(dotpaul): Remove the below suppression.
```

### PR DESCRIPTION
**Pull request by dotpaul**
Fixes #3069 - XML rules, which I will reimplement with DFA one year.

### CODE REVIEW

File: src/Microsoft.NetFramework.Analyzers/Core/Helpers/SecurityDiagnosticHelpers.cs [hunk 1]
```diff
@@ -173,9 +175,9 @@ public static bool IsExpressionEqualsIntZero(IOperation operation)
             return literal.HasConstantValue(0);
         }
 
-        private static bool IsSpecifiedProperty(IPropertySymbol symbol, INamedTypeSymbol namedType, string propertyName)
+        private static bool IsSpecifiedProperty(IPropertySymbol symbol, INamedTypeSymbol? namedType, string propertyName)
```
Comment (mavasani) [inline src/Microsoft.NetFramework.Analyzers/Core/Helpers/SecurityDiagnosticHelpers.cs]:
```suggestion
        private static bool IsSpecifiedProperty([NotNullWhen(returnValue: true)] IPropertySymbol? symbol, [NotNullWhen(returnValue: true)] INamedTypeSymbol? namedType, string propertyName)
```

File: src/Microsoft.NetFramework.Analyzers/Core/Helpers/SecurityDiagnosticHelpers.cs [hunk 2]
```diff
@@ -184,9 +186,9 @@ private static bool IsSpecifiedProperty(IPropertySymbol symbol, INamedTypeSymbol
             return false;
         }
 
-        private static bool IsSpecifiedPropertyDerived(IPropertySymbol symbol, INamedTypeSymbol namedType, string propertyName)
+        private static bool IsSpecifiedPropertyDerived(IPropertySymbol symbol, INamedTypeSymbol? namedType, string propertyName)
```
Comment (mavasani) [inline src/Microsoft.NetFramework.Analyzers/Core/Helpers/SecurityDiagnosticHelpers.cs]:
```suggestion
        private static bool IsSpecifiedPropertyDerived([NotNullWhen(returnValue: true)] IPropertySymbol? symbol, [NotNullWhen(returnValue: true)] INamedTypeSymbol? namedType, string propertyName)
```

Comment (mavasani) [reply src/Microsoft.NetFramework.Analyzers/Core/Helpers/SecurityDiagnosticHelpers.cs]: I also see a bunch of methods below which have null checks, but the parameter types are not decorated with `?`. Probably want to address those?

File: src/Microsoft.NetFramework.Analyzers/Core/Helpers/SecurityDiagnosticHelpers.cs [hunk 3]
```diff
@@ -274,7 +277,7 @@ public static string GetNonEmptyParentName(SyntaxNode current, SemanticModel mod
         /// For .NET Framework 4.X, this method returns the actual framework version instead of assembly verison of mscorlib,
         /// i.e. for .NET framework 4.5.2, this method return 4.5.2 instead of 4.0.0.0.
         /// </remarks>
-        public static Version GetDotNetFrameworkVersion(Compilation compilation)
+        public static Version? GetDotNetFrameworkVersion(Compilation compilation)
         {
             if (compilation == null || !IsTypeDeclaredInExpectedAssembly(compilation, "System.String", "mscorlib").GetValueOrDefault())
```
Comment (mavasani) [inline src/Microsoft.NetFramework.Analyzers/Core/Helpers/SecurityDiagnosticHelpers.cs]: @sharwell Do you know why compiler does not generate any warning for parameter not being declared nullable when the code has a null check up front?

File: src/Microsoft.NetFramework.Analyzers/Core/Helpers/SecurityDiagnosticHelpers.cs [hunk 4]
```diff
@@ -274,7 +277,7 @@ public static string GetNonEmptyParentName(SyntaxNode current, SemanticModel mod
         /// For .NET Framework 4.X, this method returns the actual framework version instead of assembly verison of mscorlib,
         /// i.e. for .NET framework 4.5.2, this method return 4.5.2 instead of 4.0.0.0.
         /// </remarks>
-        public static Version GetDotNetFrameworkVersion(Compilation compilation)
+        public static Version? GetDotNetFrameworkVersion(Compilation compilation)
```
Comment (mavasani) [inline src/Microsoft.NetFramework.Analyzers/Core/Helpers/SecurityDiagnosticHelpers.cs]:
```suggestion
        public static Version? GetDotNetFrameworkVersion(Compilation? compilation)
```

File: src/Tools/ReleaseNotesUtil/Program.cs [hunk 5]
```diff
@@ -71,7 +71,7 @@ private static void GetRulesJson(string nugetInstalledPackagesPath, string versi
             if (!RoslynString.IsNullOrWhiteSpace(latestRulesJsonPath))
             {
                 RuleFileContent latestContent = ReadRuleFileContent(latestRulesJsonPath);
-                Dictionary<string, RuleInfo> latestRulesById = latestContent.Rules.ToDictionary(r => r.Id);
+                Dictionary<string?, RuleInfo> latestRulesById = latestContent.Rules.ToDictionary(r => r.Id);
```
Comment (mavasani) [inline src/Tools/ReleaseNotesUtil/Program.cs]: Hmm can rule ID be null?

**Comment by dotpaul** [comment]
Thank you, @mavasani !

**Pull request merged**